pkg/terraform/errors/errors.go

```diff
@@ -16,40 +16,156 @@ limitations under the License.
 
 package errors
 
+import (
+	"fmt"
+	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/pkg/errors"
+)
+
+const (
+	levelError = "error"
+)
+
+type tfError struct {
+	logs    []byte
+	tfLogs  []*TerraformLog
+	cause   error
+	message string
+}
+
 type applyFailed struct {
-	log string
+	*tfError
+}
+
+// TerraformLog represents relevant fields of a Terraform CLI JSON-formatted log line
+type TerraformLog struct {
+	Level      string        `json:"@level"`
+	Message    string        `json:"@message"`
+	Diagnostic LogDiagnostic `json:"diagnostic"`
+}
+
+// LogDiagnostic represents relevant fields of a Terraform CLI JSON-formatted
+// log line diagnostic info
+type LogDiagnostic struct {
+	Severity string `json:"severity"`
+	Summary  string `json:"summary"`
+	Detail   string `json:"detail"`
+}
+
+func (t *tfError) Error() string {
+	if t == nil {
+		return ""
+	}
+
+	messages := make([]string, 0, len(t.tfLogs))
+	for _, l := range t.tfLogs {
+		// only use error logs
+		if l == nil || l.Level != levelError {
+			continue
+		}
+		m := l.Message
+		if l.Diagnostic.Severity == levelError && l.Diagnostic.Summary != "" {
+			m = fmt.Sprintf("%s: %s", l.Diagnostic.Summary, l.Diagnostic.Detail)
+		}
+		messages = append(messages, m)
+	}
+	return fmt.Sprintf("%s: %s", t.message, strings.Join(messages, "\n"))
+}
+
+func (t *tfError) Unwrap() error {
+	if t == nil {
+		return nil
+	}
+	return t.cause
 }
 
-func (a *applyFailed) Error() string {
-	return a.log
+func newTFError(message string, cause error, logs []byte) (string, *tfError) {
+	tfError := &tfError{
+		logs:    logs,
+		cause:   cause,
+		message: message,
+	}
+
+	tfLogs, err := parseTerraformLogs(logs)
+	if err != nil {
+		return err.Error(), tfError
+	}
+	tfError.tfLogs = tfLogs
+	return "", tfError
 }
 
-// NewApplyFailed returns a new apply failure error with given logs.
-func NewApplyFailed(log string) error {
-	return &applyFailed{log: log}
+func parseTerraformLogs(logs []byte) ([]*TerraformLog, error) {
+	logLines := strings.Split(string(logs), "\n")
+	tfLogs := make([]*TerraformLog, 0, len(logLines))
+	for _, l := range logLines {
+		log := &TerraformLog{}
+		l := strings.TrimSpace(l)
+		if l == "" {
+			continue
+		}
+		if err := jsoniter.ConfigCompatibleWithStandardLibrary.UnmarshalFromString(l, log); err != nil {
+			return nil, err
+		}
+		tfLogs = append(tfLogs, log)
+	}
+	return tfLogs, nil
+}
+
+// WrapTFError returns a new Terraform CLI failure error with given logs.
+func WrapTFError(message string, cause error, logs []byte) error {
+	if cause == nil {
+		return nil
+	}
+
+	parseError, tfError := newTFError(message, cause, logs)
+	if parseError == "" {
+		return tfError
+	}
+	return errors.WithMessage(tfError, parseError)
+}
+
+// WrapApplyFailed returns a new apply failure error with given logs.
+func WrapApplyFailed(cause error, logs []byte) error {
```

Comment:
I think _wrapping_ in std library means adding information about the upper level in the call stack, however, here we're using logs which is the actual content of the error. In fact, the error returned from `cmd.CombinedOutput()` is usually just `exited with 1` which is not really the cause.

So, I'd suggest treating this as creation of a new error instead of wrapping the command return code with logs. In fact, accepting only `logs` could be even simpler and still give us all the information we need.

```suggestion
func NewApplyFailed(cause error, logs []byte) error {
```

or 

```suggestion
func NewApplyFailed(logs []byte) error {
```

Reply:
The idea here is that we are wrapping the error returned from the `exec` library. Also because I wanted to use these two error types for apply and destroy and the generic tf error type for all operations on a workspace, we were already wrapping errors returned from the `exec` library. For instance, in the previous implementation, we were wrapping as follows in `Refresh`:
```go
	out, err := cmd.CombinedOutput()
	...
	if err != nil {
		return RefreshResult{}, errors.Wrapf(err, "cannot refresh: %s", string(out))
	}
```

Now it becomes:
```go
	out, err := cmd.CombinedOutput()
	...
	if err != nil {
		return RefreshResult{}, tferrors.WrapTFError("cannot refresh", err, out)
	}
```

And I like the idea of wrapping as much as possible as it pertains info about the context.

Reply:
Another idea here is we would like to generate a TF error only if there is an underlying error. So, converting these to wrappers is more appropriate IMHO. If other use cases arise for these error types, we can then consider adding plain constructors but our current use cases are only wrappers.

Reply:
> And I like the idea of wrapping as much as possible as it pertains info about the context.

The thing is even though it looks like that, it doesn't actually pertains any info about the context in practice; it's only about what error code is returned. I think wrapping would've made more sense if the `err` returned from `exec` actually returned with some information like output of `stderr` for example. However, what it conveys is only the fact that the execution has failed (`exited with code 1`) and we process that information with `err != nil` anyway. So even though it looks usual error wrapping, the error returned is actually misleading and we need to look at the `stderr` and/or `stdout` in all cases. If you think the error code is valuable we can preprend it to the log like `fmt.Sprintf("%s: %s", err.Error(), log)`.